pkg/build/strategies/dockerfile/dockerfile.go

```diff
@@ -0,0 +1,323 @@
+	} else {
+		// if an err on reading or opening dir, can't copy it
+		glog.V(2).Infof("Could not gather scripts from the directory %q", builder.uploadScriptsDir)
+	}
+
+	env := sti.CreateBuildEnvironment(builder.config.WorkingDir, config.Environment)
+	if len(env) > 0 {
+		buffer.WriteString("ENV")
+		for _, v := range env {
+			buffer.WriteString(fmt.Sprintf(" %s", sanitize(v)))
+		}
+		buffer.WriteString("\n")
+	}
+
+	// copy in the user's source code.
+	buffer.WriteString(fmt.Sprintf("COPY --chown=%s:0 %s %s\n", sanitize(imageUser), sanitize(builder.uploadSrcDir), sanitize(filepath.ToSlash(sourcesDestDir))))
+
+	glog.V(2).Infof("Processing secret inputs: %#v", config.Injections)
+	config.Injections = util.FixInjectionsWithRelativePath(config.ImageWorkDir, config.Injections)
+	glog.V(2).Infof("Processed secret inputs: %#v", config.Injections)
+
+	for _, injection := range config.Injections {
+		src := filepath.Join(api.Injections, injection.Source)
+		buffer.WriteString(fmt.Sprintf("COPY --chown=%s:0 %s %s\n", sanitize(imageUser), sanitize(src), sanitize(injection.Destination)))
+	}
+
+	if assembleProvided {
+		buffer.WriteString(fmt.Sprintf("RUN %s\n", sanitize(filepath.Join(scriptsDestDir, "assemble"))))
+	} else {
+		buffer.WriteString(fmt.Sprintf("# Assemble script sourced from builder image based on user input or image metadata.\n"))
+		buffer.WriteString(fmt.Sprintf("# If this file does not exist in the image, the build will fail.\n"))
+		buffer.WriteString(fmt.Sprintf("RUN %s\n", sanitize(filepath.Join(config.ImageScriptsDir, "assemble"))))
+	}
+
+	// Remove any secrets that were copied into the image,
+	// but leave configmap content alone.
+	wroteRun := false
+	for _, injection := range config.Injections {
+		if injection.Keep {
+			continue
+		}
+		if !wroteRun {
+			buffer.WriteString("RUN ")
+			wroteRun = true
+		} else {
+			buffer.WriteString(" && ")
+		}
+		buffer.WriteString(fmt.Sprintf("rm -rf %s", sanitize(injection.Destination)))
+	}
+	if wroteRun {
+		buffer.WriteString("\n")
+	}
+
+	if runProvided {
+		buffer.WriteString(fmt.Sprintf("CMD %s\n", sanitize(filepath.Join(scriptsDestDir, "run"))))
+	} else {
+		buffer.WriteString(fmt.Sprintf("# Run script sourced from builder image based on user input or image metadata.\n"))
+		buffer.WriteString(fmt.Sprintf("# If this file does not exist in the image, the build will fail.\n"))
+		buffer.WriteString(fmt.Sprintf("CMD %s\n", sanitize(filepath.Join(config.ImageScriptsDir, "run"))))
+	}
+
+	if err := builder.fs.WriteFile(filepath.Join(builder.config.AsDockerfile), buffer.Bytes()); err != nil {
+		return err
+	}
+	glog.V(2).Infof("Wrote custom Dockerfile to %s", builder.config.AsDockerfile)
+	return nil
+}
+
+// Prepare prepares the source code and tar for build.
+// NOTE: this func serves both the sti and onbuild strategies, as the OnBuild
+// struct Build func leverages the STI struct Prepare func directly below.
+func (builder *Dockerfile) Prepare(config *api.Config) error {
```

Comment:
Would this be the right place to verify that `AsDockerfile` is valid? Ex - if `AsDockerfile` is a `my/output/path/`, can we assume the user wants the output to be `my/output/path/Dockerfile` and update the `AsDockerfile` value accordingly?

Reply:
i think if additional validation logic is needed, i'd put it in build.go where we're already validating the arguments in other ways.

as for what validation/defaulting should be done with the AsDockerfile value...  yes i suppose if it ends in "/" we can append "Dockerfile" by default.